src/overlay/TCPPeer.cpp

```diff
@@ -173,105 +225,81 @@ TCPPeer::~TCPPeer()
+        // to free the OS resources.
         //
-        // But since we _might_ be dropping a hostile or unresponsive
-        // connection, we're going to just post a close() immediately after,
-        // and hope the kernel does something useful as far as putting any
-        // queued last-gasp ERROR_MSG packet on the wire.
-        //
-        // All of this is voluntary. We can also just close(2) here and be
-        // done with it, but we want to give some chance of telling peers
-        // why we're disconnecting them.
-        asio::error_code ec;
-        std::ignore = self->mSocket->next_layer().shutdown(
-            asio::ip::tcp::socket::shutdown_both, ec);
-        if (ec)
+        // It will also, at this point, cancel any pending asio
+        // read/write handlers, i.e. fire them with an error
+        // code indicating cancellation.
+        auto self_ = std::static_pointer_cast<TCPPeer>(self);
+        asio::error_code ec2;
+        self_->mSocket->close(ec2);
+        if (ec2)
         {
-            CLOG_DEBUG(Overlay, "TCPPeer::drop shutdown socket failed: {}",
-                       ec.message());
+            CLOG_DEBUG(Overlay, "TCPPeer::drop close socket failed: {}",
+                       ec2.message());
         }
-        self->mAppConnector.postOnMainThread(
-            [self]() {
-                // Close fd associated with socket. Socket is already shut
-                // down, but depending on platform (and apparently whether
-                // there was unread data when we issued shutdown()) this
-                // call might push RST onto the wire, or some other action;
-                // in any case it has to be done to free the OS resources.
-                //
-                // It will also, at this point, cancel any pending asio
-                // read/write handlers, i.e. fire them with an error code
-                // indicating cancellation.
-                asio::error_code ec2;
-                std::ignore = self->mSocket->close(ec2);
-                if (ec2)
-                {
-                    CLOG_DEBUG(Overlay, "TCPPeer::drop close socket failed: {}",
-                               ec2.message());
-                }
-            },
-            "TCPPeer: close");
-    });
+    };
+
+    maybeExecuteInBackground("TCPPeer: close", socketClose);
 }
 
 void
 TCPPeer::messageSender()
 {
     ZoneScoped;
-    releaseAssert(threadIsMain());
+    releaseAssert(!threadIsMain() || !useBackgroundThread());
 
     // if nothing to do, mark progress and return.
-    if (mWriteQueue.empty())
+    if (mThreadVars.getWriteQueue().empty())
     {
-        mWriting = false;
-        // there is nothing to send and delayed shutdown was
-        // requested - time to perform it
-        if (mDelayedShutdown)
-        {
-            shutdown();
-        }
+        mThreadVars.setWriting(false);
+        // Shutdown will be handled automatically by the timer
```

Comment:
This comment relates to a code block that was removed -- the handling of mDelayedShutdown

```suggestion
```